Added pingRead implementation.

sesh-kebab opened the pull request:
pingRead implementation as a result of discussion started on the [johnny-five issue842](https://github.com/rwaldron/johnny-five/issues/842).

Interacting with an ultrasonic sensor requires timing pulses to 10 microseconds and measuring the response to microsecond accuracy. To enable this, created a new native node plugin [raspi-sonar](https://github.com/sesh-kebab/raspi-sonar). This pull request is to add a dependency to raspi-sonar to allow implementation of the pingRead method of the [io-plugin](https://github.com/rwaldron/io-plugins) interface.

Thanks for all your help with everything and patiently answering all my questions @nebrius & @rwaldron.  Have learnt so much as result :smiley:

I'm expecting more work is needed before this PR can be accepted. But I believe the implementation has reached a stage where we can start having a conversation to move towards completing the pingRead implementation so we can start using ultrasonic sensors using jhonny-five on a RaspberryPi.

nebrius commented:
Some servo changes were just added. I suspect that the merge conflict is caused by the source mapping in the built file. Can you rebase against master and rebuild?

nebrius commented on index.js:
```diff
@@ -573,8 +574,23 @@ class Raspi extends EventEmitter {
-  pingRead() {
-    throw new Error('pingRead is not yet implemented');
+  pingRead(config, handler) {
+    const pulseOut = config.pulseOut;
```
This and the next line can be shortened to:

```JavaScript
const { pusleOut, pin } = config;
```

Yay ES6 :)

sesh-kebab replied:
I wasn't too sure myself - especially about pulseOut. I'm assuming it is used to indicate how often a reading should be taken from the ultrasonic sensor. But not sure what the unit is either. 5μs / 5ms seems too short and 5s seems too long.

rwaldron replied:
@sesh-kebab @nebrius 

It's the second arg passed to `pulseIn`

```
pulseIn(pin, HIGH);
```


> Reads a pulse (either HIGH or LOW) on a pin. For example, if value is HIGH, pulseIn() waits for the pin to go HIGH, starts timing, then waits for the pin to go LOW and stops timing. Returns the length of the pulse in microseconds or 0 if no complete pulse was received within the timeout.

(https://www.arduino.cc/en/Reference/PulseIn)

Looks like it's irrelevant to `Sonar`, so you can safely ignore it. I'll make a proper table for those properties

rwaldron replied:
Look at Particle-IO, it just ignores everything except for `pin` property: https://github.com/rwaldron/particle-io/blob/master/lib/particle.js#L544-L565

rwaldron replied:
Updated: https://github.com/rwaldron/io-plugins#pingreadsettings-handler

nebrius commented on index.js:
```diff
@@ -573,8 +574,23 @@ class Raspi extends EventEmitter {
+    
+    const pinInstance = this[getPinInstance](this.normalize(pin));
+    if (pinInstance.mode != INPUT_MODE) {
+      this.pinMode(pin, INPUT_MODE);
```
hmm...we're going to need to rethink this here. calling `pinMode` will revert the pin from using the Sonar class you implemented and instead assign it for use with the GPIO class.

I think that we should add this logic to the `pinMode` private method. I think we should create a new (internal only) mode called `PING_READ_MODE` and assign it a value of 98. Then we can call `this[pinMode](pin, PING_READ_MODE)` right here. The `new Sonar()` call will be moved to the pin mode method, although the rest of the logic will remain here. There's one catch though, we'll have to set the mode on the pinInstance to `UNKNOWN_MODE` though inside of the `pinMode` method, since Johnny-Five won't recognize `PING_READ_MODE`.

The sonar instance can be referenced here using `pinInstance.peripheral` to set up the read handler and such.

Let me know if any of this is unclear, I know it's a lot to take in.

sesh-kebab replied:
Just to confirm my understanding of all the corrections:

**New pin mode**
```javascript
const PING_READ_MODE = 98;
```

**Update pingRead to make use of the new mode***
```javascript
const pinInstance = this[getPinInstance](this.normalize(pin));
if (pinInstance.mode != PING_READ_MODE) {
  this.pinMode(pin, PING_READ_MODE);
}
```
I'm not sure about the below line though. 

>There's one catch though, we'll have to set the mode on the pinInstance to UNKNOWN_MODE though inside of the pinMode method, since Johnny-Five won't recognize PING_READ_MODE.

Based on @rwaldron 's update to [pingRead settings handler](https://github.com/rwaldron/io-plugins#pingreadsettings-handler), it read "No pin mode specified". So my question is, do we need to still worry about the catch? And if so, do you mean something like the below? (sorry if i'm completely off track here).

```javascript
pinMode(pin, mode) {
if (mode === PING_READ_MODE) //as johnny-five can't recognize PING_READ_MODE
  mode = UNKNOWN_MODE;
      
  this[pinMode]({ pin, mode });
}
```

**Use `pinInstance.peripheral` to hold the reference to an instance of Sonar**
```javascript
if (pin == LED_PIN && !(pinInstance.peripheral instanceof LED)) {
  pinInstance.peripheral = new LED();
} else {
  switch (mode) {
	case INPUT_MODE:
	  pinInstance.peripheral = new DigitalInput(config);
	  break;
	case OUTPUT_MODE:
	  pinInstance.peripheral = new DigitalOutput(config);
	  break;
	case PWM_MODE:
	case SERVO_MODE:
	  pinInstance.peripheral = new PWM(normalizedPin);
	  break;
	case PING_READ_MODE:  //<-- new code here
	  pinInstance.peripheral = new Sonar(normalizedPin);
	  break;
	default:
	  console.warn('Unknown pin mode: ' + mode);
	  break;
  }
}
```

nebrius replied:
> I'm not sure about the below line though.

> > There's one catch though, we'll have to set the mode on the pinInstance to UNKNOWN_MODE though inside of the pinMode method, since Johnny-Five won't recognize PING_READ_MODE.

> Based on @rwaldron 's update to pingRead settings handler, it read "No pin mode specified". So my question is; do we still need to worry about the catch? And if so, do you mean something like the below? (sorry if i'm completely off track here).

Hmm...ok, so here's the issue I've been thinking about: the RPi is kinda unlike the Arduino in that the "specialty" pins (I2C, UART, SPI, etc) can also be used for GPIO, so our mode logic is a little trickier on the Pi. We have to always know what mode the pin is in so we can potentially switch it to something else later. In other words, we _have_ to have a way of determining what mode a pin is it (it always has a mode, logically speaking, even if we don't specify it on the `mode` property).

We can make the logic more complicated throughout raspi-io to support pingRead not specifying pinMode, but every mode check will get more complicated and I'd really prefer to avoid that.

@rwaldron does this sound ok setting the pin mode to 98?

rwaldron replied:
I think that's fine, since Johnny-Five won't be setting a mode, you can use whatever mechanism you need to track it internally.

nebrius commented:
I left a few comments, please take a look when you get a chance, and thanks for the pull request!

sesh-kebab commented:
Thanks for all great feedback. Will definitely need time to absorb. In Tokyo for the weekend so will resume looking into starting Next week.

rwaldron commented on index.js:
```diff
@@ -573,8 +574,23 @@ class Raspi extends EventEmitter {
+    
+    const interval = setInterval(() => {
+      sonar.read(handler);
+    }, pulseOut);
```
This is incorrect—`pingRead` should only read once and emit the reading, take a look at how it's used: https://github.com/rwaldron/johnny-five/blob/master/lib/proximity.js#L192-L199

Note that particle-io only does a one time event: https://github.com/rwaldron/particle-io/blob/master/lib/particle.js#L558 and voodoospark doesn't register reporting for that pin when it receives a `PING_READ` command.

Same with Firmata.js: https://github.com/jgautier/firmata/blob/master/lib/firmata.js#L1437


`pingRead` is really just garbage and everyone should use HCSR04 sensors with an I2C backpack. Generally, once there are more than 1 of these `pingRead` operations, the process is likely being blocked _too_ much.

sesh-kebab replied:
@rwaldron the algorithm was translated from a python example. My thoughts also were that that the constant polling would tie up the CPU; the plan is to switch using interrupts in raspi-sonar once integration is completed. Would you see this as a more viable approach? Especially for using more than a single ultrasonic sensor?

rwaldron replied:
All I meant what that this call should be changed: `setInterval` => `setImmediate`

rwaldron commented on index.js:
```diff
@@ -573,8 +574,23 @@ class Raspi extends EventEmitter {
+    
+    pinInstance.peripheral.on('destroyed', () => {
+      clearInterval(interval);
+    });
```
Looks like the `"ping-read-${pin}"` event is missing as well

sesh-kebab commented:
@nebrius, @rwaldron - I've rebased and modified the code based discussion and changes requested:
* [903286e](https://github.com/sesh-kebab/raspi-io/commit/903286ede70ad222bda387251d4b7eae6af411bc)
* [5a273eb](https://github.com/sesh-kebab/raspi-io/commit/5a273eb7ce45f50ddf1a7a354981a93171ed77eb)

While I test the code out with johnny-five rather than testing raspi-io in isolation, are there other pending issues with the PR that I might have missed? Thanks again for all the help :smile:

sesh-kebab commented:
@nebrius, @rwaldron - I've rebased and modified the code based discussion and changes requested.

I've tested the code using with Johnny-five and ensured using a Proximity class is working as expected and the results (distance measurement) being produced are accurate. 

Can you please let me know if there are any other outstanding issues that I might have missed? Thanks again for all the help :smile:

rwaldron commented:
> I've tested the code with Johnny-five and ensured using a Proximity class is working as expected and the results (distance measurement) being produced are accurate.

Thank you :) that's the most important thing ❤️

@nebrius is the captain of this ship, but you get my vote!

nebrius commented:
Sorry, been busy. I have time set aside early this evening (PST) to go through this PR, so hang tight!

nebrius commented:
I'm testing things out now, although my sonar sensor seems to have died :( Not a big deal though, I should be able to work around it.

In the meantime, I opened a pair of issues on raspi-sonar for some documentation stuff, can you take a look?

sesh-kebab commented:
Yikes, sorry to hear that - I broke a sensor too while I was playing around with getting it working initially. 

I've addressed the 2 issues: [License?](https://github.com/sesh-kebab/raspi-sonar/issues/2) & [Documentation](https://github.com/sesh-kebab/raspi-sonar/issues/3) you have raised in the raspi-sonar repo. Fighting off a sore throat at the moment so was able to look at this on my day off... :mask:

nebrius commented on CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+## 5.3.0 (2016-2-19)
```
Can you try and figure out why these are showing up as a diff? It looks like GitHub thinks two of _my_ commits are new...which worries me a little.

sesh-kebab replied:
Probably something went wrong during my rebase attempt. There aren't be any changes at all to this file from my side. I'll try to roll the changes to this file back (not sure how yet).

nebrius replied:
Sorry, I somehow missed your last comment, that was my bad.

OK, I think I worked out a solution for you:

```
# Save your current work on pingread
git checkout -b temp

# Go back to master
git checkout master

# DANGER: this will delete all of your work off of master. If anything at all goes wrong, stop here and report back. Your work will be saved on branch temp.
git reset --hard f750377e

# If you've already added my repo as a remote, feel free to skip
git remote add nebrius git@github.com:nebrius/raspi-io.git

# If you already added my repo as a remote, use the name you gave it here instead of nebrius
git pull nebrius master

# This will remove your work from your master branch on GitHub and add in my latest changes
git push --force origin master

# This will be the branch that is used for the PR
git checkout -b pingread

# This will pull your branches from temp, but without my commits in the middle giving a clean history
git cherry-pick --strategy=recursive -X theirs f713bb0e 903286ed 5a273eb7 c01f2685

# Push the branch
git push origin pingread
```

After you've done those steps, open a new PR from that branch and close this one, and everything should be good!

sesh-kebab replied:
Closing this PR following the above steps and opening a new PR from pingread branch.

sesh-kebab commented:
Sorry been a bit hectic so couldn't get around to this again for a while. Created new [PR 55](https://github.com/nebrius/raspi-io/pull/55) based on the above instructions.

alexcroox commented:
Did this make it into rasp-io? Keen to start using J5 proximity with the HC-SR04 connected to Pi2

nebrius commented:
@alexcroox not yet. Check out #55 to follow along with the progress.

Also, ping @sesh-kebab, can you take a look at the feedback in #55?
